File: web/src/app/transaction/transaction_model.js
import appDetails from '_appDetails.js';
import { postData, deleteData, getData } from 'app/core/helpers/fetch.js';
import { getUserCode } from "app/core/authentication/authentication.js"
import { downloadFile } from "app/core/helpers/file_handler.js";
import isEmpty from "app/core/helpers/is_empty.js";

const model = {
    TransactionKey: 0,
    MemberKey: 0,
    Amount: 0,
    Category: "",
    IsApproved: "",
    ApprovedDate: "",
    CreatedBy: "",
    CreatedDate: null,
    ModifiedBy: "",
    ModifiedDate: null
};

const prepData = (detailID, category, data) => {
    var detail = _.clone(model);

    // Use only for transactions
    if (category !== null) {
        detail.MemberKey = detailID;
        detail.Category = category;
        detail.TypeOfLoan = data.TypeOfLoan;
        detail.StartDueDate = data.StartDueDate;
        detail.CreatedBy = getUserCode();
        detail.CreatedDate = new Date();
        detail.ModifiedBy = getUserCode();
        detail.ModifiedDate = new Date();
    }

    // Shared for Download Computation and New Loan
    detail.Amount = data.Amount;
    detail.Interest = data.Interest;
    detail.Term = data.Term;
    detail.ServiceFee = data.ServiceFee;
    detail.InsuranceAmount = data.InsuranceAmount;
    detail.FixedDepositAmount = data.FixedDepositAmount;
    detail.DocumentationAmount = data.DocumentationAmount;
    detail.SavingsDepositAmount = data.SavingsDepositAmount;
    detail.BalancePreviousLoanAmount = data.BalancePreviousLoanAmount;
    detail.InterestPreviousLoanAmount = data.InterestPreviousLoanAmount;
    detail.DepositSavings = data.DepositSavings;
    detail.DepositShareCapitalAmount = data.DepositShareCapitalAmount;
    return detail;
};

const prepPayment = (detailID, data) => {
    var detail = _.clone(model);
    detail.MemberKey = detailID;
    detail.LoanKey = data.loanID;
    detail.Category = "Payment";
    detail.Amount = data.Amount;
    detail.Interest = data.Interest;
    detail.Principal = data.Principal;
    detail.Penalty = data.Penalty;
    detail.CreatedBy = getUserCode();
    detail.CreatedDate = new Date();
    detail.ModifiedBy = getUserCode();
    detail.ModifiedDate = new Date();
    return detail;
};

const addPayment = (detailID, dataDetail) => {
    var detail = prepPayment(detailID, dataDetail);
    return postData(`transaction/add/payment`, detail).then((data) => {
        if (data && data.ok) {
            return data.json();
        } else {
            return false;
        }
    });
};

const addTransaction = (detailID, category, dataDetail) => {
    var detail = prepData(detailID, category, dataDetail);
    return postData(`transaction/add/${category}`, detail).then((data) => {
        if (data && data.ok) {
            return data.json();
        } else {
            return false;
        }
    });
};

const deleteTransaction = (transactionKey, category) => {
    return deleteData(`transaction/delete/${category}/${transactionKey}`).then((data) => {
        if (data && data.ok) {
            return data.json();
        } else {
            return false;
        }
    });
};

const getMemberTransactionList = (memberKey, type, filters) => {
    const pageCount = !isEmpty(filters.pageCount) && filters.pageCount;
    const page = isEmpty(filters.page) ? 1 : filters.page;
    const filterBy = isEmpty(filters.filterByValue) ? "" : filters.filterByValue;
    const searchBy = isEmpty(filters.searchValue) ? "" : filters.searchValue;
    return getData(`transaction/${type}/list/${memberKey}?page=${page}&pageCount=${pageCount}&filterBy=${filterBy}&searchBy=${searchBy}`)
        .then(data => data.json())
};

const getLoanTransactionList = (loanID, type, filters) => {
    const pageCount = !isEmpty(filters.pageCount) && filters.pageCount;
    const page = isEmpty(filters.page) ? 1 : filters.page;
    const filterBy = isEmpty(filters.filterByValue) ? "" : filters.filterByValue;
    const searchBy = isEmpty(filters.searchValue) ? "" : filters.searchValue;

    return getData(`transaction/${type}/list/${loanID}?page=${page}&pageCount=${pageCount}&filterBy=${filterBy}&searchBy=${searchBy}`)
        .then(data => data.json())
};

const getMembersWithTransaction = (filters) => {
    const pageCount = !isEmpty(filters.pageCount) && filters.pageCount;
    const page = isEmpty(filters.page) ? 1 : filters.page;
    const filterBy = isEmpty(filters.filterByValue) ? "" : filters.filterByValue;
    const searchBy = isEmpty(filters.searchValue) ? "" : filters.searchValue;
    return getData(`transaction/list?page=${page}&pageCount=${pageCount}&filterBy=${filterBy}&searchBy=${searchBy}`)
        .then(data => data.json())
};

const getLoanPaymentDetails = (loanID) => {
    return getData(`transaction/loan/details/${loanID}`)
        .then(data => data.json())
};

const getTypeOfLoans = () => {
    return getData('transaction/typeofloans/list')
        .then(data => data.json())
};

const downloadComputation = (memberKey, transactData) => {
    let fileName = "";
    var detail = prepData(memberKey, null, transactData);
    return postData(`transaction/calculation?memberKey=${memberKey}&amount=${transactData.Amount}`, detail)
        .then(response => {
            if (!response.ok) {
                throw new Error('Network response was not OK');
            }
            let contentDisposition = response.headers.get("Content-Disposition");
            fileName = contentDisposition.substring(
                contentDisposition.indexOf("filename=\"") + 10,
                contentDisposition.lastIndexOf(".xlsx\";")
            );;
            return response.blob();
        })
        .then(blob => {
            downloadFile(blob, `Computation - ${fileName}.xlsx`);
            return fileName;
        });
};

export {
    model,
    getMemberTransactionList,
    getLoanTransactionList,
    addTransaction,
    addPayment,
    deleteTransaction,
    getMembersWithTransaction,
    downloadComputation,
    getLoanPaymentDetails,
    getTypeOfLoans
};

File: server/coploan/coploan/Models/TransactionModel.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace coploan.Models
{
    public class TransactionDetails
    {
        public int TransactionKey { get; set; }
        public int MemberKey { get; set; }
        public decimal Amount { get; set; }
        public string CreatedBy { get; set; }
        public DateTime CreatedDate { get; set; }
        public string ModifiedBy { get; set; }
        public DateTime ModifiedDate { get; set; }

    }

    public class LoanDetails: TransactionDetails
    {
        public DateTime StartDueDate { get; set; }
        public decimal Interest { get; set; }
        public int Term { get; set; }
        public string TypeOfLoan { get; set; }
        public decimal ServiceFee { get; set; }
        public decimal InsuranceAmount { get; set; }
        public decimal FixedDepositAmount { get; set; }
        public decimal DocumentationAmount { get; set; }
        public decimal SavingsDepositAmount { get; set; }
        public decimal BalancePreviousLoanAmount { get; set; }
        public decimal InterestPreviousLoanAmount { get; set; }
    }

    public class LoanDetailsBalance: LoanDetails
    {
        public decimal Balance { get; set; }
    }

    public class InterestDetails : TransactionDetails { }

    public class DepositDetails : TransactionDetails 
    { 
        public decimal DepositSavings { get; set; }
        public decimal DepositShareCapitalAmount { get; set; }
    }

    public class PaymentDetails : TransactionDetails 
    { 
        public int LoanKey { get; set; }
        public decimal Interest { get; set; }
        public decimal Principal { get; set; }
        public decimal Penalty { get; set; }
    }

    public class LoanPaymentDetails : TransactionDetails
    {
        public decimal PaidAmount { get; set; }
        public decimal UnpaidAmount { get; set; }
    }

    public class LoanComputation
    {
        public decimal Amount { get; set; }
        public int Term { get; set; }
        public decimal Payment { get; set; }
        public decimal Interest { get; set; }
        public decimal Principal { get; set; }
        public decimal Balance { get; set; }
        public decimal ServiceFee { get; set; }
        public decimal InsuranceAmount { get; set; }
        public decimal FixedDepositAmount { get; set; }
        public decimal DocumentationAmount { get; set; }
        public decimal SavingsDepositAmount { get; set; }
        public decimal BalancePreviousLoanAmount { get; set; }
        public decimal InterestPreviousLoanAmount { get; set; }

    }

    public class ComputedMonthlyLoan
    {
        public int Month { get; set; }
        public decimal Payment { get; set; }
        public decimal Interest { get; set; }
        public decimal Principal { get; set; }
        public decimal Balance { get; set; }
    }
}
